### ISSUE TITLE
B&W G3 locks up at NT Text Setup

### ISSUE THREAD

**Issue by nugz4life** [Wack0/maciNTosh#17]
So the issue i am facing is the machine locks up on the part where its loading the cdrom file system.
My setup is as follows hard drive as a master and CDROM as a slave on the secondary bus. 
I have let the system sit for a good 10 to 30 minutes  and no progress.

**Comment by Wack0** [comment]
This is a hang in early NT kernel init or HAL init. What HAL did you load?

**Comment by nugz4life** [comment]
The G3 Powermac one

**Comment by Wack0** [comment]
What CPU does your yosemite have?

**Comment by Dandumontp** [comment]
I had the same problem and by booting with a single memory stick (128 MB) it worked.

**Comment by G-Nug85** [comment]
Confirmed on my rev-A B&W the setup works fine with just 128MB of ram single module. Anything else fails

**Comment by Wack0** [comment]
Do you have mismatched sizes of RAM sticks in your yosemite? I might be setting up an incorrect memory map in that case.

**Comment by Dandumontp** [comment]
Yep. The initial  configuration was 256 + 128 (the G3 Blue is very tricky with old RAM)

**Comment by Wack0** [comment]
I'd like a dump of the memory map in open firmware:

```
dev /memory
.properties
```
If there's a hole in the memory map when mismatched RAM sizes are used, I'm definitely setting the memory map up wrong.

**Comment by Dandumontp** [comment]
I will try that today.

**Comment by Wack0** [comment]
I found and fixed a couple of issues when calculating physical memory size, so try this test build:

[nt_arcfw_grackle_fwonly_test20240728_1626.zip](https://github.com/user-attachments/files/16404204/nt_arcfw_grackle_fwonly_test20240728_1626.zip)

**Comment by G-Nug85** [comment]
OK i will give this a try thanks

**Comment by Wack0** [comment]
I was doing some tests with the usb stack earlier, might have accidentally left it enabled non-working in that build, so try this one:
[nt_arcfw_grackle_fwonly_test20240728_1934.zip](https://github.com/user-attachments/files/16404863/nt_arcfw_grackle_fwonly_test20240728_1934.zip)

**Comment by Wack0** [comment]
Another person with this issue discovered it was related to bad RAM, for what it's worth: https://forums.macrumors.com/threads/nt-on-powerpc-macs.2431186/post-33286633

**Comment by TechandMusic462** [comment]
Just to add, I had the issue getting stuck at "Loading CD-ROM Filesystem" on my iMac G3 Trayloader when trying to troubleshoot the HAL getting stuck (which ended up being the now-solved CUDA driver issue), when I tried changing the 512MB RAM config (2x256MB) to just 256MB (1 stick). Either stick would cause it to freeze there, both are fine. No issues on Mac OS 9 in any config. It did seem more picky about memory, basically.

**Comment by Wack0** [comment]
Yeah, I fixed some issues in stage1 recently related to memory map stuff.

For the record, I have 512MB in my Lombard, and never had any issue at kernel init time.

**Comment by b-** [comment]
Just want to mention that I had this issue on my PDQ/Wallstreet 2, which had 2x 256 MB sticks in it. Removing one stick got me through setup, and putting it back afterwards seems fine. But Windows does _not_ see more than 128 MB of RAM, regardless of whether there's one or both of the 256 MB sticks installed.

I haven't reinstalled Mac OS 9 yet, so I'm not sure I can get to the OF prompt yet to check memory properties, but I'll get you that when I can.

**Comment by Wack0** [comment]
NT itself uses the memory map that ARC firmware sets up, which uses physical memory size that stage1 gives it. For Old World specifically, this is `Gestalt(gestaltPhysicalRAMSize)` - perhaps the toolbox ROM this early in boot has issues with RAM above 128MB.

You can get to OF prompt by holding cmd+option+O+F at the chime.

**Comment by b-** [comment]
Thanks, I didn't think you could get to the OF prompt on any Old World machines via keyboard. Turns out you need to use the right option key for some reason on this model, but I got there. 

![image](https://github.com/user-attachments/assets/5d3f819d-e480-4291-b115-7a9c656d7844)

**Comment by b-** [comment]
Also I have to take back what I said about it booting to a graphical installed NT 4.0 with all 512 MB installed -- I think that was a fluke, because I haven't gotten it to boot to an NT desktop with all 512 MB since...

It's indeed possible that it didn't work with all the memory in OS 9, I don't remember ever checking. I'm burning a Mac OS 9 disc now...

**Comment by b-** [comment]
So, after reinstalling Mac OS 9.2.2 dual-boot, I can report that the 256 MB RAM sticks I have are showing up as 128 MB each in _About This Computer_.

I think this might be because they're double-sided, or something? I'm not familiar with how this works… but nevertheless 256 MB seems usable under Mac OS 9 but isn't detected by the ARC FW.
